File: 0101-0200/0132-Palindrome Partitioning II/0132-Palindrome Partitioning II.cpp
class Solution {
public:
    int minCut(string s) {
        int n = s.size();
        if (n == 0) {
            return 0;
        }
        
        vector<int> dp(n);
        vector<vector<bool>> palindrome(n, vector<bool>(n, false));
        for (int j = 0; j < n; ++j) {
            dp[j] = j;
            for (int i = j; i >= 0; --i) {
                if (s[i] == s[j] && (j - i < 2 || palindrome[i + 1][j - 1])) {
                    palindrome[i][j] = true;
                    if (i == 0) {
                        dp[j] = 0;
                    }
                    else {
                        dp[j] = min(dp[j], 1 + dp[i - 1]);
                    }
                }
            }
        }
        
        return dp[n - 1];
    }
};

File: 0101-0200/0132-Palindrome Partitioning II/0132-Palindrome Partitioning II.py
class Solution:
    def minCut(self, s: str) -> int:
        dp = [i - 1 for i in range(len(s) + 1)]
        for i in range(len(s)):
            j = 0
            while i - j >= 0 and i + j < len(s) and s[i - j] == s[i + j]:
                dp[i + j + 1] = min(dp[i + j + 1], 1 + dp[i - j])
                j += 1
                
            j = 1
            while i - j + 1 >= 0 and i + j < len(s) and s[i - j + 1] == s[i + j]:
                dp[i + j + 1] = min(dp[i + j + 1], 1 + dp[i - j + 1])
                j += 1
        return dp[-1]
